server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Scanner.java

```diff
@@ -38,100 +40,126 @@
   private ScanDataSource isolatedDataSource;
   private boolean sawException = false;
   private boolean scanClosed = false;
+  protected Semaphore scannerSemaphore;
 
   Scanner(Tablet tablet, Range range, ScanOptions options) {
     this.tablet = tablet;
     this.range = range;
     this.options = options;
+    scannerSemaphore = new Semaphore(1, true);
   }
 
-  public synchronized ScanBatch read() throws IOException, TabletClosedException {
-
-    if (sawException)
-      throw new IllegalStateException("Tried to use scanner after exception occurred.");
+  public ScanBatch read() throws IOException, TabletClosedException {
 
     if (scanClosed)
```

Comment:
I made a typo when writing the patch from my other machine. I made the boolean volatile there but in talking with you I will move the scanClosed below the acquisition to avoid the issue entirely and ensure cache coherency.